_posts/2024-07-24-jenkins-agents.md

```diff
@@ -0,0 +1,304 @@
+    + [SSH Agents](#sshagents)
+        + [Configuring the controller as code to implement the agents](#cascode)
+        + [Deploying the agents](#dagents)
+            + [The Docker agent](#dagent)
+            + [The VM agent](#vmagent)
+    + [The Docker Plugin (`docker-plugin`)](#dplugin)
+        + [Configuring Docker as a Cloud provider for kworkflow's Jenkins CI](#cdc)
+        + [Preparing the environment](#pte)
+        + [Creating the Cloud](#ctc)
++ [The integration tests](#tint)
++ [Final thoughts](#ft)
++ [References](#resources)
+
+During my GSoC project of planning a Jenkins CI for kworkflow, I spent plenty of time experimenting with different possibilities of agent types for the infrastructure.
+This post aims to register every attempt of agent implementation so far, their respective implementation guide for a Jenkins server, and my insights about them for the kworkflow context.
+
+## What is a Jenkins agent? <a name="what" />
+
+Jenkins is designed to be a **distributed** automation server, which means that it is best used when the workload is balanced between different computing nodes. These computing nodes are the Jenkins agents and represent
+computing environments assigned to execute pipeline steps.
+
+The use of agents to run jobs exclusively is also recommended as a safer approach, as it gives appropriate controller isolation. Let's remember that a build in kworkflow's Jenkins will execute code from contributors, which
+may be broken or malicious. Therefore, one should avoid exposing the controller node directly to such code and disable the controller's built-in node.
+
+## Static agents vs. Cloud <a name="vs" />
+
+In Jenkins, there are two primary ways to provide agents: You can either configure and launch them statically, by configuring your own containers and VMs and accessing them via (outbound) SSH connections or 
+(inbound) TCP connections, or dynamically provide them through a Cloud service, such as Amazon EC2 instances, Kubernetes Pods, etc.
+
+## What I've tried so far <a name="sofar" />
+
+### The Docker Pipeline Plugin (`docker-workflow`) <a name="dworkflow" />
+
+My first attempt at configuring an agent for Jenkins was by using the Docker Pipeline Plugin. It allows one to define in the **jenkinsfile** the Docker image that will be run for a specific pipeline/stage and
+dynamically use it in the pipeline when required.
+
+It is pretty simple to use. You just have to install the plugin for your Jenkins server and write a jenkinsfile invoking the dynamic docker container. You can pull a container image from the docker hub
+defining 
+
+```
+agent {
+    docker { image 'chosen-image' }
+}
+```
+
+or even using a custom Dockerfile from the root of the project with
+
+```
+agent {
+    dockerfile true 
+}
+```
+
+but notice you need to define a SCM for your Pipeline or use a Multibranch Pipeline to use the custom Dockerfile since it needs a repository to get the Dockerfile from.
```

Comment:
By "SCM", do you mean Source Control Management? Maybe use the full name with the acronym in parenthesis.